binderhub/build_local.py

```diff
@@ -0,0 +1,227 @@
+"""
+Contains build of a docker image from a git repository.
+"""
+
+import asyncio
+from collections import defaultdict
+import datetime
+from functools import partial
+import json
+import os
+import threading
+from urllib.parse import urlparse
+import subprocess
+import warnings
+from tornado import ioloop
+
+from tornado.ioloop import IOLoop
+from tornado.log import app_log
+
+from .utils import rendezvous_rank, KUBE_REQUEST_TIMEOUT
+from .build import ProgressEvent, Build
+
+
+# https://github.com/jupyterhub/repo2docker/blob/2021.08.0/repo2docker/utils.py#L13-L58
+def _execute_cmd(cmd, capture=False, **kwargs):
+    """
+    Call given command, yielding output line by line if capture=True.
+
+    Must be yielded from.
+    """
+    if capture:
+        kwargs["stdout"] = subprocess.PIPE
+        kwargs["stderr"] = subprocess.STDOUT
+
+    proc = subprocess.Popen(cmd, **kwargs)
```

Comment:
Should we use `asyncio.subprocess` instead?

Reply:
I'm not sure. `build.submit` is run in a `ThreadPoolExecutor`
https://github.com/jupyterhub/binderhub/blob/be94c46009c970617be45e27c9a6ed95203b4af0/binderhub/app.py#L714-L715
https://github.com/jupyterhub/binderhub/blob/be94c46009c970617be45e27c9a6ed95203b4af0/binderhub/builder.py#L427
So does that mean `asyncio.subprocess` wouldn't do anything?

Reply:
Ah yeah ok - if it's run in a threadpool it can't be async.